File: app/src/main/java/com/linderaredux/view/Alphabetik.java
package com.linderaredux.view;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Typeface;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.linderaredux.R;

import java.util.Arrays;

import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

public class Alphabetik extends RecyclerView {

    //Default Alphabet
    private String[] alphabet = {"A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K",
            "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z", "#"};

    private int width;
    private float fontSize;
    private int selectedFontSize;
    private static int itemsColor;
    private int selectedItemColor;
    private int selectedItemBackground;

    private SectionIndexAdapter adapter;
    private LinearLayoutManager linearLayoutManager;

    public Alphabetik(Context context, AttributeSet attrs) {
        super(context, attrs);
        this.setOverScrollMode(OVER_SCROLL_NEVER);
        getAttributes(context, attrs);
        initRecyclerView();
    }

    private void getAttributes(Context context, AttributeSet attrs) {
        TypedArray ta = context.obtainStyledAttributes(attrs, R.styleable.Alphabetik);

        width = ta.getDimensionPixelSize(R.styleable.Alphabetik_width, 15);

        int defaultSize = (int) spToPixel(context, 12);
        int attFontSizeValue = ta.getDimensionPixelSize(R.styleable.Alphabetik_fontSize, defaultSize);
        fontSize = pixelsToSp(context, attFontSizeValue);

        int aItemsColor = R.styleable.Alphabetik_itemsColor;
        if (ta.hasValue(R.styleable.Alphabetik_itemsColor)) {
            itemsColor = getColor(ta.getResourceId(aItemsColor, 0));
        }

        //Recycle
        ta.recycle();
    }

    private int getColor(int id) {
        return ContextCompat.getColor(getContext(), id);
    }

    private float pixelsToSp(Context context, int px) {
        float scaledDensity = context.getResources().getDisplayMetrics().scaledDensity;
        return px / scaledDensity;
    }

    private float spToPixel(Context context, int sp) {
        float scaledDensity = context.getResources().getDisplayMetrics().scaledDensity;
        return sp * scaledDensity;
    }

    @SuppressLint("WrongConstant")
    private void initRecyclerView() {
        adapter = new SectionIndexAdapter(alphabet, getContext());
        linearLayoutManager = new LinearLayoutManager(getContext(), LinearLayoutManager.VERTICAL, false);
        this.setHasFixedSize(true);
        this.setAdapter(adapter);
        this.setLayoutManager(linearLayoutManager);
    }

    public void setAlphabet(String[] alphabet) {
        Arrays.sort(alphabet);
        this.alphabet = alphabet;
        initRecyclerView();
    }

    public void onSectionIndexClickListener(SectionIndexClickListener sectionIndexClickListener) {
        adapter.onSectionIndexClickListener(sectionIndexClickListener);
    }

    private String[] generateAlphabet() {
        String[] alphabetTemp = new String[27];
        int index = 0;
        for (char c = 'A'; c <= 'Z'; c++) {
            alphabetTemp[index] = "" + c;
            index++;
        }
        return alphabetTemp;
    }

    public void setLetterToBold(String letter) {
        int index = Arrays.asList(alphabet).indexOf(letter);
        String regex = "[0-9]+";
        if (letter.matches(regex)) {
            index = alphabet.length - 1;
        }
        adapter.setBoldPosition(index);
        this.linearLayoutManager.scrollToPositionWithOffset(index, 0);
        this.getAdapter().notifyDataSetChanged();
    }

    class SectionIndexAdapter extends Adapter<SectionIndexAdapter.ViewHolder> {

        private int boldPosition = 0;
        private String[] alphabet;
        private LayoutInflater mInflater;
        private SectionIndexClickListener sectionIndexClickListener;

        SectionIndexAdapter(String[] alphabet, Context context) {
            this.alphabet = alphabet;
            this.mInflater = LayoutInflater.from(context);
        }

        //LISTENER
        public void onSectionIndexClickListener(SectionIndexClickListener sectionIndexClickListener) {
            this.sectionIndexClickListener = sectionIndexClickListener;
        }

        public void setBoldPosition(int position) {
            this.boldPosition = position;
        }

        @Override
        public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            View view = mInflater.inflate(R.layout.item_letter, parent, false);
            return new ViewHolder(view);
        }

        @SuppressLint("ResourceAsColor")
        @Override
        public void onBindViewHolder(ViewHolder holder, int position) {
            String letter = alphabet[position];
            holder.tvLetter.setText(letter);

            //Set current position to bold
            Typeface normalTypeface = Typeface.defaultFromStyle(Typeface.NORMAL);
            Typeface boldTypeface = Typeface.defaultFromStyle(Typeface.BOLD);
            holder.tvLetter.setTypeface(position == boldPosition ? boldTypeface : normalTypeface);
            holder.tvLetter.setTextColor(position == boldPosition ? ContextCompat.getColor(getContext(), R.color.black) : ContextCompat.getColor(getContext(), R.color.medium_grey));

            //Custom Font size
            holder.tvLetter.setTextSize(TypedValue.COMPLEX_UNIT_SP, fontSize);

            //Custom color
            if (itemsColor != 0) {
                holder.tvLetter.setTextColor(itemsColor);
            }
        }

        @Override
        public int getItemCount() {
            return alphabet.length;
        }

        //VIEW HOLDER
        class ViewHolder extends RecyclerView.ViewHolder implements OnClickListener {

            private TextView tvLetter;

            public ViewHolder(View itemView) {
                super(itemView);
                tvLetter = (TextView) itemView.findViewById(R.id.tvLetter);
                itemView.setOnClickListener(this);
            }

            @Override
            public void onClick(View view) {
                if (sectionIndexClickListener != null) {
                    String character = "" + tvLetter.getText().toString();
                    sectionIndexClickListener.onItemClick(view, this.getPosition(), character);
                    setLetterToBold(character);
                }
            }
        }
    }

    //INTERFACES
    public interface SectionIndexClickListener {
        void onItemClick(View view, int position, String character);
    }
}

File: app/src/main/java/com/linderaredux/view/Alphabetik.kt
package com.linderaredux.view

import android.annotation.SuppressLint
import android.content.Context
import android.content.res.TypedArray
import android.graphics.Typeface
import android.util.AttributeSet
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView

import com.linderaredux.R

import java.util.Arrays

import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView

class Alphabetik(context: Context, attrs: AttributeSet) : RecyclerView(context, attrs) {

    private var alphabet = arrayOf("A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z", "#")

    private var fontSize: Float = 0.toFloat()
    private val selectedFontSize: Int = 0
    private val selectedItemColor: Int = 0
    private val selectedItemBackground: Int = 0

    private var adapter: SectionIndexAdapter? = null
    private var linearLayoutManager: LinearLayoutManager? = null

    init {
        this.overScrollMode = View.OVER_SCROLL_NEVER
        getAttributes(context, attrs)
        initRecyclerView()
    }

    private fun getAttributes(context: Context, attrs: AttributeSet) {
        val ta = context.obtainStyledAttributes(attrs, R.styleable.Alphabetik)

        val defaultSize = spToPixel(context, 12).toInt()
        val attFontSizeValue = ta.getDimensionPixelSize(R.styleable.Alphabetik_fontSize, defaultSize)
        fontSize = pixelsToSp(context, attFontSizeValue)

        val aItemsColor = R.styleable.Alphabetik_itemsColor
        if (ta.hasValue(R.styleable.Alphabetik_itemsColor)) {
            itemsColor = getColor(ta.getResourceId(aItemsColor, 0))
        }

        ta.recycle()
    }

    private fun getColor(id: Int): Int {
        return ContextCompat.getColor(context, id)
    }

    private fun pixelsToSp(context: Context, px: Int): Float {
        val scaledDensity = context.resources.displayMetrics.scaledDensity
        return px / scaledDensity
    }

    private fun spToPixel(context: Context, sp: Int): Float {
        val scaledDensity = context.resources.displayMetrics.scaledDensity
        return sp * scaledDensity
    }

    @SuppressLint("WrongConstant")
    private fun initRecyclerView() {
        adapter = SectionIndexAdapter(alphabet, context)
        linearLayoutManager = LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)
        this.setHasFixedSize(true)
        this.setAdapter(adapter)
        this.layoutManager = linearLayoutManager
    }

    fun setAlphabet(alphabet: Array<String>) {
        Arrays.sort(alphabet)
        this.alphabet = alphabet
        initRecyclerView()
    }

    fun onSectionIndexClickListener(sectionIndexClickListener: SectionIndexClickListener) {
        adapter!!.onSectionIndexClickListener(sectionIndexClickListener)
    }

    private fun generateAlphabet(): Array<String?> {
        val alphabetTemp = arrayOfNulls<String>(27)
        var index = 0
        var c = 'A'
        while (c <= 'Z') {
            alphabetTemp[index] = "" + c
            index++
            c++
        }
        return alphabetTemp
    }

    fun setLetterToBold(letter: String) {
        var index = Arrays.asList(*alphabet).indexOf(letter)
        val regex = "[0-9]+"
        if (letter.matches(regex.toRegex())) {
            index = alphabet.size - 1
        }
        adapter!!.setBoldPosition(index)
        this.linearLayoutManager!!.scrollToPositionWithOffset(index, 0)
        this.getAdapter()!!.notifyDataSetChanged()
    }

    internal inner class SectionIndexAdapter(private val alphabet: Array<String>, context: Context) : RecyclerView.Adapter<SectionIndexAdapter.ViewHolder>() {

        private var boldPosition = 0
        private val mInflater: LayoutInflater = LayoutInflater.from(context)
        private var sectionIndexClickListener: SectionIndexClickListener? = null

        fun onSectionIndexClickListener(sectionIndexClickListener: SectionIndexClickListener) {
            this.sectionIndexClickListener = sectionIndexClickListener
        }

        fun setBoldPosition(position: Int) {
            this.boldPosition = position
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            val view = mInflater.inflate(R.layout.item_letter, parent, false)
            return ViewHolder(view)
        }

        @SuppressLint("ResourceAsColor")
        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            val letter = alphabet[position]
            holder.tvLetter.text = letter

            //Set current position to bold
            val normalTypeface = Typeface.defaultFromStyle(Typeface.NORMAL)
            val boldTypeface = Typeface.defaultFromStyle(Typeface.BOLD)
            holder.tvLetter.typeface = if (position == boldPosition) boldTypeface else normalTypeface
            holder.tvLetter.setTextColor(if (position == boldPosition) ContextCompat.getColor(context, R.color.black) else ContextCompat.getColor(context, R.color.medium_grey))

            //Custom Font size
            holder.tvLetter.setTextSize(TypedValue.COMPLEX_UNIT_SP, fontSize)

            //Custom color
            if (itemsColor != 0) {
                holder.tvLetter.setTextColor(itemsColor)
            }
        }

        override fun getItemCount(): Int {
            return alphabet.size
        }

        internal inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {

            val tvLetter: TextView = itemView.findViewById(R.id.tvLetter) as TextView

            init {
                itemView.setOnClickListener(this)
            }

            override fun onClick(view: View) {
                if (sectionIndexClickListener != null) {
                    val character = "" + tvLetter.text.toString()
                    sectionIndexClickListener!!.onItemClick(view, this.position, character)
                    setLetterToBold(character)
                }
            }
        }
    }

    interface SectionIndexClickListener {
        fun onItemClick(view: View, position: Int, character: String)
    }

    companion object {
        private var itemsColor: Int = 0
    }
}